File: quicksort/hoares/hoares.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int partition(int* list, int low, int high);
void quickSort(int* list, int low, int high);
void swap(int* a, int* b);

int main(){
    int arr[] = {5,7,4,3,6,8,9,1,2,10};

    printf("before sort: ");
    for(int i = 0; i < 10; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");

    quickSort(arr, 0, 9);

    printf("after sort: ");
    for(int i = 0; i < 10; i++){
        printf("%d ", arr[i]);
    }
}

void quickSort(int* list, int low, int high){
    if(low < high){
        int pivot = partition(list, low, high);

        quickSort(list, low, pivot);
        quickSort(list, pivot + 1, high);
    }
}

int partition(int* list, int low, int high){
    int pivot = list[(high+low)/2]; // take middle element as the pivot

    int i = low - 1;
    int j = high + 1;
    
    while(1){
        while(i < high && list[++i] < pivot);
        while(j > low && list[--j] > pivot);
        if(i >= j) return j;
        swap(&list[i], &list[j]);
    }
}

void swap(int* a, int* b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

File: quicksort/hoares/hoares.py
def QuickSort(array, low, high):
    if low < high:
        pivot = partition(array, low, high)     # find pivot
        QuickSort(array, low, pivot)          
        QuickSort(array, pivot+1, high)         

def partition(array, low, high):
    pivot = array[(low+high)/2] # make pivot the middle of the array
    
    i = low - 1
    j = high + 1
    
    while(True):
        i += 1
        while array[i] < pivot:
            i += 1
        j -= 1
        while array[j] > pivot:
            j -= 1
        
        if i >= j:
            return j
        array[i], array[j] = array[j], array[i]     # swap

test_list = [10,9,4,5,3,2,6,7,8,1]
QuickSort(test_list, 0, len(test_list)-1)
print(test_list)

